shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -360,78 +427,190 @@
+          // and the platform view rect.
+          joined_rect.intersect(platform_view_rect);
+          // Clip the background canvas, so it doesn't contain any of the pixels drawn
+          // on the overlay layer.
+          background_canvas->clipRect(joined_rect, SkClipOp::kDifference);
+          // Get a new host layer.
+          auto layer = GetLayer(gr_context,                //
+                                ios_context,               //
+                                picture,                   //
+                                joined_rect,               //
+                                current_platform_view_id,  //
+                                overlay_count              //
+          );
+          did_submit &= layer->did_submit_last_frame;
+          platform_view_layers[current_platform_view_id].push_back(layer);
+          overlay_count++;
+        }
+      }
     }
-  }
-  picture_recorders_.clear();
-  if (composition_order_ == active_composition_order_) {
-    composition_order_.clear();
-    return did_submit;
-  }
-  DetachUnusedLayers();
-  active_composition_order_.clear();
-  UIView* flutter_view = flutter_view_.get();
+    background_canvas->drawPicture(picture);
+  }
+  // If a layer was allocated in the previous frame, but it's not used in the current frame,
+  // then it can be removed from the scene.
+  RemoveUnusedLayers();
+  // Organize the layers by their z indexes.
+  BringLayersIntoView(platform_view_layers);
+  // Mark all layers as available, so they can be used in the next frame.
+  layer_pool_->RecycleLayers();
+  // Reset the composition order, so next frame starts empty.
+  composition_order_.clear();
+
+  return did_submit;
+}
 
+void FlutterPlatformViewsController::BringLayersIntoView(LayersMap layer_map) {
+  auto flutter_view = flutter_view_.get();
+  auto zIndex = 0;
   for (size_t i = 0; i < composition_order_.size(); i++) {
-    int view_id = composition_order_[i];
-    // We added a chain of super views to the platform view to handle clipping.
-    // The `platform_view_root` is the view at the top of the chain which is a direct subview of the
-    // `FlutterView`.
-    UIView* platform_view_root = root_views_[view_id].get();
-    UIView* overlay = overlays_[view_id]->overlay_view;
-    FML_CHECK(platform_view_root.superview == overlay.superview);
-    if (platform_view_root.superview == flutter_view) {
-      [flutter_view bringSubviewToFront:platform_view_root];
-      [flutter_view bringSubviewToFront:overlay];
-    } else {
+    auto platform_view_id = composition_order_[i];
+    auto layers = layer_map[platform_view_id];
+    auto platform_view_root = root_views_[platform_view_id].get();
+
+    if (platform_view_root.superview != flutter_view) {
       [flutter_view addSubview:platform_view_root];
-      [flutter_view addSubview:overlay];
-      overlay.frame = flutter_view.bounds;
+    } else {
+      platform_view_root.layer.zPosition = zIndex++;
```

Comment:
Just curious, what's the difference between zIndex and `bringSubviewToFront`

Reply:
bringSubviewToFront also controls the zPosition. However, it's more limited since it increments `zPosition` to the highest value possible.